API - first draft

gunzip opened the pull request.

cloudify commented:
A couple of quick feedbacks:

* let's stick to lowercase endpoints
* we need to add versioning (e.g. `/v1` prefix)

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+        "application/json"
+    ],
+    "paths": {
+        "/Messages/{id}": {
```
Why `Messages` if we're talking about Notifications? :)

gunzip replied:
IMHO 'messages' it's a broader concept that includes notifications (but not only)

gunzip replied:
About lowercase endpoints, just want to point out that SCIM endpoints are capitalized:
http://www.simplecloud.info/

cloudify replied:
the majority of API endpoints are lowercase, avoid mixing cases makes the endpoint more predictable

cloudify replied:
also here we're talking specifically about notifications, we should avoid ambiguity and try to be consistent, messages is indeed a broader concept that implies a conversation - in our case the communication is unidirectional, so it's notifications

dgreco replied:
Why not using the Zalando REST APIs policies? they have put on line a document with nice suggestions for defining REST APIs using swagger/openapi

cloudify replied:
Thanks @dgreco will check them out!

gunzip replied:
[Zalando REST APIs policies](http://zalando.github.io/restful-api-guidelines/json-guidelines/JsonGuidelines.html) are very good indeed but also VERY opinionated. 

ie. there are some directives like "**Must**: Do Not Use URI Versioning" which IMHO are hard to be applied blindly.

Moreover, **if** we want to be compliant and coherent with [SCIM API specs](https://tools.ietf.org/html/rfc7644) we have to start with a different style (ie. camelCased endpoints and so on).

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                ],
+                "responses": {
+                    "200": {
+                        "description": "",
```
we need a description, like `Success: message exist`

gunzip replied:
that would work if there's only one (the same) message for every HTTP status code, which is not always true.

cloudify replied:
isn't this the description of a successful GET message request?

gunzip replied:
gotcha, it seems you cannot see (and therefore, edit) this field from spotlight ui :)
I've added the description manually to the specs

cloudify replied:
I don't see the descriptions in the PR

cloudify replied:
👍

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                            "$ref": "#/definitions/message-output"
+                        }
+                    },
+                    "400": {
```
we need a description of each error

gunzip replied:
see comment above

cloudify replied:
I don't understand, I'm saying we need a description for each status code, see [here](https://swagger.io/docs/specification/describing-responses/) for an example, in the *HTTP Status Codes* section

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    }
+                }
+            },
+            "patch": {
```
what's the meaning of updating a notification? one of the principles we had was "immutability"

gunzip replied:
updating the status (ie. from pending to sent)

cloudify replied:
I don't understand, who's gonna call this endpoint? it's the notification system that will (internally) update the state of the notification, the clients will be able to read the state, not update it...

Care making an example use case?

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    }
+                }
+            },
+            "delete": {
```
I would skip this for now since the semantic is not immediately clear

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                }
+            }
+        },
+        "/Messages": {
```
This is unclear to me, is it supposed to return all messages sent by anybody to anybody else?

cloudify replied:
also I would leave out all search criteria, not clear to me what's the use case of this endpoint in the context of the early adopters

gunzip replied:
ie: retrieve messages by reference_id

cloudify replied:
all the specified parameters are optional, meaning that one can call `/Messages` without any filter, what's the expected output?

gunzip replied:
the commentend line is the endpoint for the sending operation (it's a POST actually)

cloudify replied:
sorry got confused, I thought it was a GET endpoint :)

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                        "$ref": "#/parameters/trait:pagination:page_size"
+                    },
+                    {
+                        "$ref": "#/parameters/trait:messageSearchQueryString:title"
```
out of scope

cloudify replied:
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "description": "One fiscal code (CF).\n",
+                    "type": "string"
+                },
+                "from": {
```
from should be implicit, provided by the authentication process

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                "from": {
+                    "type": "string"
+                },
+                "client_id": {
```
what is this?

gunzip replied:
sender (oauth client) external id (from another service)

cloudify replied:
we don't have oauth, so?

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "type": "string",
+                    "format": "uuid"
+                },
+                "reference_id": {
```
I would live this out for now, let's keep the API as simple as possible

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "type": "string",
+                    "description": "An optional identifier you generate if you don’t want to use notify’s id. It can be used to identify a single notification or a batch of notifications."
+                },
+                "reply_to_message_id": {
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "type": "string",
+                    "format": "uuid"
+                },
+                "thread_id": {
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "type": "string",
+                    "format": "uuid"
+                },
+                "priority": {
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "description": "This parameter specifies how long (in seconds) the message should be kept in storage if the device is offline. The maximum time to live supported is 4 weeks, and the default value is 4 weeks.",
+                    "minimum": 0
+                },
+                "expires_at": {
```
out of scope for now

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "type": "string",
+                    "format": "date-time"
+                },
+                "tracking": {
```
what is this?

gunzip replied:
to turn on tracking (ie. link tracking on emails or number of opened emails and so on)

cloudify replied:
definitely out of scope :)

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                "tracking": {
+                    "type": "string"
+                },
+                "tags": {
```
really?

gunzip replied:
this are message topics (user can opt out)

cloudify replied:
let's keep all these "meta features" out for now

cloudify replied:
this about the simples possible notification API you can imagine

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                        "type": "string"
+                    }
+                },
+                "store_content": {
```
out of scope for now

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                    "properties": {
+                        "mail": {
+                            "type": "object",
+                            "properties": {
```
these are out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                        "$ref": "#/parameters/trait:messageSearchQueryString:to"
+                    },
+                    {
+                        "$ref": "#/parameters/trait:messageSearchQueryString:from"
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                        "$ref": "#/parameters/trait:messageSearchQueryString:from"
+                    },
+                    {
+                        "$ref": "#/parameters/trait:messageSearchQueryString:in"
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                        "$ref": "#/parameters/trait:messageSearchQueryString:in"
+                    },
+                    {
+                        "$ref": "#/parameters/trait:messageSearchQueryString:unread"
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                        "$ref": "#/parameters/trait:messageSearchQueryString:unread"
+                    },
+                    {
+                        "$ref": "#/parameters/trait:messageSearchQueryString:starred"
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                        "$ref": "#/parameters/trait:messageSearchQueryString:starred"
+                    },
+                    {
+                        "$ref": "#/parameters/trait:messageSearchQueryString:received_before"
```
all these are out of scope, I would keep only the pagination

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                ],
+                "parameters": [
+                    {
+                        "name": "body",
```
I would call it `message` or `notification` since `body` has a very specific meaning (the content of a message)

gunzip replied:
'body' it's a swagger spec special keyword in this context (the POST body) and cannot be changed :)

cloudify replied:
I'm actually referring to line 173, `"name": "body"`

gunzip replied:
still applies :)

cloudify replied:
I can't find this constraint in the spec, `name` is the name of the field, it's value can be anything

gunzip replied:
Right. there's seem there's no way in the stoplight UI to change this value.

There are some other issues with the stoplight UI (like: it loses the manual customizations, like the status messages for responses)

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                "tags": [
+                    "Messages"
+                ],
+                "parameters": [
```
keep only pagination for now

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+                }
+            }
+        },
+        "/Client/{id}/messages": {
```
what's this for?

gunzip replied:
all messages sent by a client (sender)

cloudify replied:
I would scope admin endpoints under an `/admin` prefix

also, let's leave admin stuff for later and focus on the customers (senders)

gunzip replied:
that's for customers' (sender) back-office, to list sent messages (not for admins). 
otherwise customers have to retrieve messages one by one.

cloudify replied:
let's leave this out of scope for now and limit the spec to the core use cases related to sending notifications, no admin/backoffice stuff

gunzip replied:
still disagree. it's an useful feature even for a test-demo envinroment (retrieve sent messages without storing the id)

cloudify replied:
I'm not saying it's not useful, nor that we're not gonna add it later, I'm just saying that for this first MVP we should leave it out as it's not part of the core use case (sending a notification)

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+            "in": "query",
+            "type": "integer",
+            "minimum": 0,
+            "default": 1
```
default should be the first page `0` ?

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+            "minimum": 0,
+            "default": 1
+        },
+        "trait:pagination:page_size": {
```
I would use a fixed page size for now, we should give to the client as few decisions to make as possible

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                    "minLength": 2,
+                    "maxLength": 2
+                },
+                "data": {
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                    "type": "string",
+                    "faker": "lorem.paragraph"
+                },
+                "language": {
```
out of scope

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                    "description": "This parameter specifies how long (in seconds) the message should be kept in storage if the device is offline. The maximum time to live supported is 4 weeks, and the default value is 4 weeks.",
+                    "minimum": 0
+                },
+                "tags": {
```
out of scope (for now), we need first to gather more knowledge from the senders

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+            "properties": {
+                "status": {
+                    "type": "string",
+                    "enum": [
```
we need an explanation for each one of these states

gunzip replied:
added in description field

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                    "type": "string",
+                    "enum": [
+                        "PENDING",
+                        "SENDING",
```
what's the difference between pending and sending?

gunzip replied:
pending means queued (indeed it's better to call it this way probably), sending means 'delivered to the sending service' (ie. sms gateway)

cloudify replied:
I don't like `QUEUED` and `SENDING`, they both leak system abstractions (that notifications can be possibly queued and that they get relayed to a 3rd party system).

Also, we should use an [open-ended list of states](https://zalando.github.io/restful-api-guidelines/compatibility/Compatibility.html#should-used-openended-list-of-values-xextensibleenum-instead-of-enumerations).

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                        "PENDING",
+                        "SENDING",
+                        "SENT",
+                        "DELIVERED",
```
we can't provide DELIVERED nor READ for now

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                },
+                "error_code": {
+                    "type": "string",
+                    "enum": [
```
what are the meaning of these error_codes and when they're supposed to be returned?

gunzip replied:
when the status is set to FAILED (by the backend service) error code and message fields are populated with the reason that caused the delivery to fail.

cloudify replied:
I don't see how we can differentiate between permanent failure and temporary failure. What happens when a notification is temporary failed?

gunzip replied:
a retry ? =)

cloudify replied:
from the point of view of the API is still a permanent failure, the API doesn't ha ve a "retry" action

the client MAY decide to retry by calling again the send notification endpoint, thus getting back a new notification ID

gunzip replied:
I didn't mean the client has to retry. It's an information for the client that there was an error during the delivery but the backend will retry to send the message (causing some delay). Anyway, we can just leave this field open.

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                        "TEMPORARY_FAILURE"
+                    ]
+                },
+                "service_id": {
```
what is this?

gunzip replied:
that's the id of the sending service (ie. sms gateway). we can get rid of this for now.

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+        "message-status": {
+            "title": "Message Status",
+            "type": "object",
+            "properties": {
```
we need a description for each property

gunzip replied:
added in the description field of the specs

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                }
+            }
+        },
+        "message-output": {
```
this looks more like an `AcceptedMessage` or something like that, give to a type a name that describes what is it

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,504 @@
+                            "type": "string",
+                            "format": "date-time"
+                        },
+                        "updated_at": {
```
notifications cannot be updated

gunzip replied:
what about notification status ?

cloudify replied:
that's an internal status, it's not something that gets updated by an API client

gunzip replied:
right, but it's an information for the client to read (ie. when the status has changed last time ?)

cloudify replied:
that would be the latest of `queued_at`, `sent_at`, etc...

cloudify commented on users.json:
```diff
@@ -0,0 +1,1423 @@
+{
```
sorry but I can't make up my mind for why we have a 1,423 lines of spec for a simple preferences API, SCIM may be an inspiration to us but we can't just take all of it

gunzip replied:
now it's about 500 lines.

gunzip replied:
we have to decide what is our 'externalId'

that could be the fiscal number (which is not unique) or the SPID id (unique but there could be more than one for each user)

cloudify replied:
is the SPID ID globally unique? I thought it was unique in the scope of a single IdP

gunzip replied:
right. we can stick to userName = fiscalNumber and get rid of externalId.

cloudify commented:
I suggest we move to YAML specs, I find them much easier to read than JSON

gunzip commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+            ],
+            "get": {
+                "operationId": "listMessagesByUser",
+                "summary": "List Messages (by user)",
```
stated in "notes"

cloudify replied:
👍

mccalv replied:
be careful with ID, sounds like a numeric id, naturally prone to Insecure Direct Object References. 
I would use UUID always, this at least limit the attacker area and comes handy with the logs.

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+        "https"
+    ],
+    "paths": {
+        "/Messages/{id}": {
```
still called Messages, still capitalized

gunzip replied:
this is in coherence with the SCIM API (which is camelCase with capitalized endpoints). 

Should we chose a different style for the notifications API ?

cloudify replied:
ok, but make it `Notifications` then

gunzip replied:
I think 'Messages', being broader, is more future proof. There's nothing, technically speaking, that prevents the API being used as a two-way means of communication.

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                    "name": "id",
+                    "in": "path",
+                    "required": true,
+                    "type": "string"
```
I remember below a notification ID is a UUID

gunzip replied:
yes

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                        }
+                    },
+                    "400": {
+                        "$ref": "#/responses/trait:apiErrors:400"
```
We need descriptions of return codes at the request level, not in the trait. The meaning of a 404 changes from one endpoint to another.

For instance what's the meaning of 400 returned by this method? I don't see any way a GET for a specific notification could return a Bad Request

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                }
+            }
+        },
+        "/Users/{id}/messages": {
```
lowercase

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                        "$ref": "#/parameters/trait:pagination:page"
+                    },
+                    {
+                        "$ref": "#/parameters/trait:pagination:page_size"
```
I would leave this out for now, we can have a fixed page size

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+            "default": 1,
+            "minimum": 1
+        },
+        "trait:pagination:page_size": {
```
remove

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+            "description": "",
+            "schema": {
+                "type": "object",
+                "description": "Object Not Found.",
```
this should me more specific and depends on the called function

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                }
+            }
+        },
+        "trait:apiErrors:429": {
```
let's leave this out for now, we haven't defined the specs of how quota should work yet

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                        "READ",
+                        "FAILED"
+                    ],
+                    "description": "QUEUED: the message has been accepeted for delivery\n\nSENDING: the message is in charge of the delivery service (ie. sms gatewy or mail server)\n\nSENT: the message was sent by the delivery service\n\nDELIVERED: the message reached the user inbox (ie. mail server or phone)\n\nREAD: the user read the message\n\nFAILED: something wong happened, check the error c ...
```
not all channels support all states, this may be confusing and should be throughly explained

I would limit the states to 

`QUEUED`: accepted for delivery

`DELIVERED`: delivered to the channel

`FAILED`: delivery failed

`TIMED_OUT`: in case a TTL was set

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                "time_to_live": {
+                    "type": "integer",
+                    "description": "This parameter specifies how long (in seconds) the message should be kept in storage if the device is offline. The maximum time to live supported is 4 weeks, and the default value is 4 weeks.",
+                    "minimum": 0
```
not a vary sane minimum, I would say, the TTL should be at least 1h and less than 1 year

gunzip replied:
these values are taken from the firebase cloud messaging API, obviously we can change them.

maybe 1 year it's too much.

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                        "TEMPORARY_FAILURE"
+                    ]
+                },
+                "sent_at": {
```
later

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+                    "type": "string",
+                    "format": "date-time"
+                },
+                "read_at": {
```
later

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,478 @@
+            "title": "MessageStatus",
+            "type": "object",
+            "properties": {
+                "status": {
```
`accepted_at` ?

gunzip replied:
queued_at (to reflect state names)

cloudify replied:
👍

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+    "swagger": "2.0",
+    "info": {
+        "version": "v1 - demo",
+        "title": "SCIM",
```
these should describe the preferences API, with a not that the API is a subset of SCIM

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+        "https"
+    ],
+    "paths": {
+        "/Users/{userKey}": {
```
we should use the same names for the parameters in both APIs for consistency

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+    "paths": {
+        "/Users/{userKey}": {
+            "parameters": [
+                {
```
this should be the same definition used by the notification API - actually I believe we can refactor common definitions in shared definitions (e.g. a user ID)

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                "tags": [
+                    "Users"
+                ],
+                "description": "Queries user identity in the organization domain.",
```
we need better descriptions here, that explains the preferences API

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                        "description": "Client is not sufficiently authorized"
+                    },
+                    "403": {
+                        "description": "Invalid token passed"
```
also error messages should be consistent between the APIs

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                    }
+                }
+            },
+            "patch": {
```
not needed for the early adopters, they're going to be read only initially

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                    }
+                }
+            },
+            "put": {
```
not needed for the early adopters, they're going to be read only initially

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                }
+            }
+        },
+        "/Users": {
```
this will be an admin feature, leave it out for now

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                    }
+                }
+            },
+            "post": {
```
this will be an admin feature, leave it out for now

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                        "fiscalNumber": {
+                            "type": "string"
+                        },
+                        "optOut": {
```
later

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                        "$ref": "#/definitions/Im"
+                    }
+                },
+                "externalId": {
```
not needed

gunzip replied:
this could be the fiscal code number (see other comment above)

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                        "$ref": "#/definitions/PhoneNumber"
+                    }
+                },
+                "ims": {
```
out of scope for now

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                        "$ref": "#/definitions/Email"
+                    }
+                },
+                "phoneNumbers": {
```
out of scope for now

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                    "description": "The user's display name",
+                    "faker": "internet.userName"
+                },
+                "name": {
```
not needed, any SPID based service will already have this information

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+            "description": "Describes a new user",
+            "type": "object",
+            "properties": {
+                "displayName": {
```
not needed

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                "preferredLanguage": {
+                    "type": "string"
+                },
+                "urn:scim:schemas:extension:italia:1.0": {
```
I would just use the `fiscalNumber` as the `id` and/or `userName`

cloudify commented on users.json:
```diff
@@ -0,0 +1,533 @@
+                "totalResults"
+            ]
+        },
+        "UserFullName": {
```
also not needed

cloudify commented on notifications.json:
```diff
@@ -0,0 +1,861 @@
+{
+    "swagger": "2.0",
+    "info": {
+        "version": "1.0",
```
`1.0.0`

gunzip commented:
I've Updated the API specs according to comments and uploaded the YAML files. Please add any other comment to there starting a review (so we can 'close' comments as well).

Some consideration:

1. Probably common errors (ie. 403 unauthorized) should not be documented *inside* the specs, but on a wiki / textual doc. This is coherent with https://zalando.github.io/restful-api-guidelines/http/Http.html where they states that "In most cases it's not useful to document technical errors that are not in control of the service provider unless the status code convey application-specific semantics. The list of status code that can be omitted from API specifications includes but is not limited to... [status codes follow]".

2. I've added the problem+json model to the message status (to carry error codes and descriptions), as stated in https://zalando.github.io/restful-api-guidelines/common-data-objects/CommonDataObjects.html#must-use-common-error-return-objects. This is a "naive" usage as tipically the problem object should be the root object returned by the http response and not nested inside the payload.

3. We must take a decision about API style. We can keep SCIM API apart (leaving them camelCased) and use a totally different style for notifications API (zalando-like). I favour this approach.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+  title: "Notification API"
+  description: "Notification API."
+host: "localhost"
+basePath: "/api/v1"
```
`/api` is not needed

gunzip replied:
To put the "/api" prefix is a suggestion that comes from Zalando guidelines: https://zalando.github.io/restful-api-guidelines/naming/Naming.html

This may be useful in case you have private API aside public ones.

cloudify replied:
agreed, as we will have `/admin` APIs

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+  /Messages/{id}: 
+    parameters: 
+      - 
+        name: "id"
```
I think it's worth being more specific with the name, to avoid confusion and ambiguity, like `notificationId`

gunzip replied:
According to Zalando's guidelines "id" *must* be the standard name for an identifier:

https://zalando.github.io/restful-api-guidelines/common-data-objects/CommonDataObjects.html

cloudify replied:
ok

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      summary: "Get Message"
+      tags: 
+        - "Messages"
+      description: "Retrieve a message already submitted for delivery."
```
"Returns a previously submitted message."

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      tags: 
+        - "Messages"
+      description: "Retrieve a message already submitted for delivery."
+      responses: 
```
we also need 404

gunzip replied:
see comment about common errors.

cloudify replied:
404 is not a common error (see my comment below) but is action-specific (only gets) and has special meaning to each endpoint (user does not exist, message does not exist, etc...)

common errors are like authentication errors (same error and meaning for all endpoints)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+  /Users/{id}/Messages: 
+    parameters: 
+      - 
+        name: "id"
```
this is `userId`

gunzip replied:
same as above

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+        in: "path"
+        required: true
+        type: "string"
+        format: "uuid"
```
this is a fiscal number

gunzip replied:
I'm thinking about this. The fiscal number is discoverable. Probably the use case here is to get messages only for the logged in user (the one that shares his uuid with the service provider).

gunzip replied:
it's a matter of id discoverability. service provider knows the user's uuid only for users that subscribed the service itself while the fiscal number is a public parameter and opens the possibility for the service provider to query every user message queue.

cloudify replied:
> it's a matter of id discoverability (enumerability). service provider knows the user's uuid only for users that subscribed the service itself 

I don't understand how the service provider could get the UUID of the user since from the SPID auth he only gets the fiscal number, so we need to provide a way for the service provider to map the fiscal number to a UUID, so the problem persists (the service provider can enumerate all fiscal numbers and get their associated UUIDs)  

> while the fiscal number is a public parameter and opens the possibility for the service provider to query every user message queue.

but only trusted service providers will be able to retrieve a user messages (i.e. our app backend), nobody else

unless you intend to have users authorize the service provider to access their data through an oAuth flow, bu that we decided is out of scope for now

please explain

gunzip replied:
I was thinking of oauth flow (and/or some "/User/Me" endpoint used to retrieve the current user's data without having to know the uuid in advance). But I agree it's an edge case and we can use the CF instead. This is one of the changes that assumes we are diverging from the SCIM specs. I'm still not convinced this is the more convenient path but let's start with this piece :)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+        format: "uuid"
+    get: 
+      operationId: "listMessagesByUser"
+      summary: "List Messages by user"
```
Returns all the submitted messages the have the user as destination.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+    get: 
+      operationId: "listMessagesByUser"
+      summary: "List Messages by user"
+      tags: 
```
remove tags, let's think about this later, unless this is needed to satisfy our core use cases

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      summary: "List Messages by user"
+      tags: 
+        - "Messages"
+      description: "Retrive all messages for this user. \n\nMessages will be returned in inverse delivery order (from last to first)"
```
Returns all the submitted messages the have the user as destination.
Messages will be returned in inverse submission order (from last to first).

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      description: "Retrive all messages for this user. \n\nMessages will be returned in inverse delivery order (from last to first)"
+      responses: 
+        200: 
+          description: "Found."
```
this is tricky, what happens when I request the messages for a user that doesn't have any messages? Still 200 Found?

gunzip replied:
404 (see comment about common errors)

gunzip replied:
https://stackoverflow.com/questions/13366730/proper-rest-response-for-empty-table

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+            $ref: "#/definitions/MessageCollection"
+  /Messages: 
+    post: 
+      operationId: "sendMessage"
```
I would stick with `submit` prefix (we used it in the description and other parts of the spec)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+    post: 
+      operationId: "sendMessage"
+      summary: "Send Message"
+      tags: 
```
remove tags for now

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      summary: "Send Message"
+      tags: 
+        - "Messages"
+      description: "Submit a message for delivery."
```
Submits a message for delivery to a user.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+        - "Messages"
+      description: "Submit a message for delivery."
+      parameters: 
+        - 
```
`name` should be on this line

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+          schema: 
+            $ref: "#/definitions/MessageDefinition"
+      responses: 
+        202: 
```
Since we are returning a UUID of a newly created message, I would go for `201 Created`, with a `Location` pointing to `/Messages/UUID`

cloudify replied:
202 assumes the request is asynchronous and will be completed later (i.e. the message will be created later)

in our case the creation is synchronous since we're returning the UUID

gunzip replied:
right

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      timed_out_at: 
+        type: "string"
+        format: "date-time"
+      problem: 
```
what is this?

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+    properties: 
+      title: 
+        type: "string"
+        minLength: 1
```
do we really want to allow titles with just one character?

gunzip replied:
almost every channel has a minimum of one. why not ? (what's a suitable minimum ?)

cloudify replied:
the minimum is there for catching errors, I can't think about a meaningful text of 1 character, 2 neither, maybe 3

gunzip replied:
ok

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      title: 
+        type: "string"
+        minLength: 1
+        maxLength: 255
```
why 255?

gunzip replied:
it fits in a SMS

cloudify replied:
[looks like it's less than that](https://stackoverflow.com/questions/21788876/what-is-the-maximum-sms-message-length)

also here you're saying that `title` should be the text-only version of the message provided in `body`, not really the subject of the email? so `title` should also contain the message

this is the problem I've described above, `title` and `body` go together and they're not necessary redundant (the whole email message is made of title+body), saying that the system can decide to deliver just the content of `title` without its `body`, gives it a different meaning and requires the `title` to contain a shorter version of what's described in the `body`

gunzip replied:
ok

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+  MessageContent: 
+    title: "MessageContent"
+    type: "object"
+    properties: 
```
I think we should not assume a certain structure of the message (title/body), as only email has this structure

Let's start very simple and have just `body_short` as plain text message, with a certain limit - it could make sense to [max to ~100 characters](https://stackoverflow.com/questions/6307748/what-is-the-maximum-length-of-a-push-notification-alert-text)

gunzip replied:
right but the email *needs* this structure (if you want to send some formatting), otherwise you're stuck with plain text and/or messages without an object (or content).

if you stick to only one field (body_short) is that the content of the email or the object or both ?

cloudify replied:
this is a discussion we already had - we have to chose one of these two options:

1. having the sender explicitly decide the destination channel and providing a message formatted for that specific channel (i.e. subject and body for email, short text for sms, etc...)
2. hiding the channel decision in the system, making it based on user preferences, hiding it from the sender, having some logic that takes the message and formats it for the channel

since `postMessage` doesn't allow setting a channel, I'm assuming we're going for option 2

in this second option, we decided to hide the channels from the sender, so we can either:

1. set for the amount of information that satisfies the "simplest" channel (e.g. SMS) and upgrade this content for richer channels (e.g. email) - this is fairly simple to do (e.g. display a simple pre-formatted text in an email) 
2. set for the the amount of information that satisfies the "richest" channel (e.g. HTML email) and downgrade this content for simpler channels (e.g SMS) - this is harder to do (how to you fit an HTML email in an SMS?) - an imperfect option would be to add indirection (e.g. send a link by SMS pointing to the HTML email)
3. option in between, ask for different formats of the same message without leaking the underlying abstractions (i.e. we may add/remove channels in the future, so we don't want to talk about specific channels like email, sms or push notifications, instead we want to talk about different formats for the same content, think about a kind of responsive version of the message, e.g. short text or rich long text, etc...)

gunzip replied:
ok let me recap. 'title' becomes 'body_short' and 'body_html' becomes 'body_long' ?

cloudify replied:
yes, I would say so, let's leave HTML formatting out for now, in the future we may have `body_markdown` or something like that

gunzip replied:
ok

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+    type: "object"
+    properties: 
+      to: 
+        description: "One fiscal code (CF).\n"
```
The fiscal code of the destination user.

Btw we need also to specify a pattern wherever we ask for a fiscal code.

gunzip replied:
what about `[A-Z]{6}[0-9]{2}[ABCDEHLMPRST][0-9]{2}[A-Z][0-9]{3}[A-Z]`

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+        type: "string"
+      dry_run: 
+        type: "boolean"
+        description: "This parameter, when set to true, allows developers to test a request without actually sending a message.\n\nThe default value is false."
```
we need to be more detailed here in the semantics, what happens? is the message created but not delivered?

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+        description: "This parameter specifies how long (in seconds) the message should be kept in storage if the device is offline. The maximum time to live supported is 1 year, and the default value is 1 hour."
+        minimum: 3600
+        maximum: 31536000
+      tags: 
```
no tags for now

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      status: 
+        type: "string"
+        description: "ACCEPTED: the message has been accepted for delivery\n\nDELIVERED: the message was delivered to the channel\n\nFAILED: something wong happened, check the error code and message\n\nTIMED_OUT:  in case a TTL was set\n"
+        enum: 
```
use `x-extensible-enum`

gunzip replied:
we can but this will prevent any code generator to apply any validation (or to generate a client that can handle the actual enum)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+    properties: 
+      status: 
+        type: "string"
+        description: "ACCEPTED: the message has been accepted for delivery\n\nDELIVERED: the message was delivered to the channel\n\nFAILED: something wong happened, check the error code and message\n\nTIMED_OUT:  in case a TTL was set\n"
```
ACCEPTED: the message has been accepted for delivery but not yet delivered.

    DELIVERED: the message has been delivered to the destination

as with delivered, we need more specific here, what happens when multiple channels are configured?
perhaps we should say that has it been delivered to at least one channel
finally, `delivered` may be misleading as we're just relaying to a channel, we don't know if it has actually been delivered or not

    FAILED: the message delivery failed 

    TIMED_OUT: the message was not delivered in time

TTL is always set (there's a default)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+      - 
+        type: "object"
+        properties: 
+          id: 
```
`messageId`

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+                $ref: "#/definitions/MessageStatus"
+              mobile: 
+                $ref: "#/definitions/MessageStatus"
+              web: 
```
what is `web` ?

gunzip replied:
web notification api (for PWA app)

cloudify replied:
so we have:

1. email
2. mobile via sms (?)
3. mobile via push notification to the app
4. mobile / desktop via push notification to the browser

so web is really browser ?

gunzip replied:
yes, we can refactor channel names: 'sms', 'push', 'browser', 'email'

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+          status: 
+            type: "object"
+            default: {}
+            properties: 
```
not sure that the sender needs to know on which channels the message has been received and when, we should not expose the information (at least at this stage) and just tell the sender that the message has been delivered to at least a channel

here you can see the ambiguity between "message" and "notification": 

if we assume that this system provides an "inbox" for the user, the message *always* gets delivered to the user inbox, so as long as sendMessage succeeds, the sender knows the message has been delivered to the user inbox

on the other end, a message being delivered to one's inbox should trigger a notification (hey you got a new message), notifications can contain a preview of the message and gets delivered over multiple channels and multiple times (in case one wants an acknowledge)

I feel like we haven't explored enough the complexity of this kind of interaction (messages vs notifications) - as we want this API to do both (provide an inbox for messages and deliver notifications)

gunzip replied:
there are cases when the inbox (permanent storage) should be skipped and this is a decision in charge of the sender.

cloudify replied:
that's my point, that exception is caused by the ambiguity of the API, is this a message (must be stored in the inbox for later retrieval) or just an ephemeral notification (storing not required) ?

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+    title: "MessageCollection"
+    type: "object"
+    properties: 
+      resources: 
```
`messages` ?

gunzip replied:
but it's not generic. it's an object containing Message(s). 'items' would be wrong, 'messages' too similar to 'message' maybe.

cloudify replied:
either you are specific or you're generic - btw in the [Zalando guidelines](https://zalando.github.io/restful-api-guidelines/pagination/Pagination.html) they use `items`

gunzip replied:
it's ok to use "items" in the response but in this case MessageCollection is the name of the model in the swagger specs (never exposed to the caller of the API)

cloudify replied:
I was referring to `resources` that is a property exposed to the caller, right?

gunzip replied:
Sorry, I thought you were referring to 'MessageCollection', now I get it :)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+        default: []
+        items: 
+          $ref: "#/definitions/Message"
+      totalResults: 
```
this could be expensive to calculate and I don't see which use case needs this information, we could just have a `next` item linking to the next page

gunzip replied:
we can use HATEOAS semantic like described here: https://zalando.github.io/restful-api-guidelines/pagination/Pagination.html (so we can use 'items'). 

totalResults field was included in coherence with the SCIM API that returns list of resources, but we can switch to HATEOAS here.

cloudify replied:
yes

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,220 @@
+          description: "Message found."
+          schema: 
+            $ref: "#/definitions/Message"
+  /Users/{id}/Messages: 
```
a cursor parameter is needed

gunzip replied:
there where the pagination parameters we got rid of. we can get them back.

gunzip replied:
(as cursors navigation is not possible with uuid, we are stuck with pagination params)

gunzip replied:
i don't get it. what's 'zzzzz' exactly in this contest ?

cloudify replied:
`zzzzz` is the value of the cursor (a pointer), it's just an example, can be anything, a number, a string, it's opaque to the caller

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+info: 
+  version: "0.0.1"
+  title: "Preferences"
+  description: "The API lets you manage attributes associated with users (user preferences).\nYou can get any attribute for any user (read only) according to the rights you owe on the user resource. \nThere rights are set up at moment you've subscribed the service exposed by this API."
```
"This API provides access to user preferences."

Documentation about what is available is better described in each method that has permission based behavior.

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+  license: 
+    name: "MIT"
+host: "localhost"
+basePath: "/api/v1"
```
`/api` is not needed

gunzip replied:
same as above

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+schemes: 
+  - "https"
+paths: 
+  /Users/{id}: 
```
`userId`

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+paths: 
+  /Users/{id}: 
+    parameters: 
+      - 
```
no newline after `-`

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+      - 
+        name: "id"
+        in: "path"
+        description: "The key of the user to query. The user must be in the organization domain"
```
The fiscal number of the user.

gunzip replied:
This is the uuid in the SCIM API.

if we want to retrieve users by an attribute which is not the uuid (ie. the userName = fiscal Number) we probably better stick to the SCIM spec for filtering: 
http://www.simplecloud.info/specs/draft-scim-api-01.html

cloudify replied:
this is one case where we're trying to fit our use cases in the SCIM specs, we identify users by fiscal code (also in the notifications API), we should do the same here

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+        description: "The key of the user to query. The user must be in the organization domain"
+        required: true
+        type: "string"
+        format: "uuid"
```
this should be a fiscal number

gunzip replied:
see comment above

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+    get: 
+      operationId: "getUser"
+      summary: "Get User"
+      tags: 
```
remove tags

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+      summary: "Get User"
+      tags: 
+        - "Users"
+      description: "Retrieve one user's preferences (ie. email and/or preferred language)."
```
Returns a user's preferences

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+      description: "Retrieve one user's preferences (ie. email and/or preferred language)."
+      responses: 
+        200: 
+          description: "The request has succeeded."
```
The user was found

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+      tags: 
+        - "Users"
+      description: "Retrieve one user's preferences (ie. email and/or preferred language)."
+      responses: 
```
+ 404

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+          schema: 
+            $ref: "#/definitions/User"
+definitions: 
+  UserDefinition: 
```
if this is already under `definitions` why the suffix `Definition` ?

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+        description: "Indicates the User's preferred written or spoken language. Generally used for selecting a localized User interface. Valid values are concatenation of the ISO 639-1 two letter language code, an underscore, and the ISO 3166-1 2 letter country code; e.g., 'en_US' specifies the language English and country US."
+    required: 
+      - "userName"
+  User: 
```
this is confusing, there a `User` definition that contains a `UserDefinition` definition?

gunzip replied:
it's the same for Message and MessageDefinition. There's a common pattern: to describe a model which is the input (without id and other properties) and the same model + some attributes (ie. the id) for the output (the stored value). 

We have to choose a naming convention here (ie. UserInput + UserOutput)

cloudify replied:
I don't understand why we need a nested definition, a user is ID + attributes, what's the problem with that?

cloudify replied:
i.e.

```
GET /Users/123
{
  "id": "123",
  "email": "...",
  ...
}

gunzip replied:
When you do a POST you don't usually provide an `id` (so you have a simpler / naked model without some attributes that is used as an *input* in the request). To DRY it's good to extend this simpler model to use the complex one in the responses, otherwise you have to cut&paste the entire definition and keep in sync the input model with the output model inside the specs:

https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#composition-and-inheritance-polymorphism

cloudify replied:
for what I can see, having a model with all the attributes but the ID (provided in the URI), solves all our needs, doesn't it?

gunzip replied:
indeed, you have a model with the id and another model without it. but the presence (or absence) of the id does not usually cover all cases.

for example the message output has the 'status' field which is not present in the input message model. and you have two calls (getMessage and getMessagesByUser) that return the message output. in this case having an output model is useful (rather than hardcoding the id and other fields directly in the response structure).

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+    description: "Describes a user"
+    type: "object"
+    properties: 
+      userName: 
```
we already have one ID (`User#userId`), what is this?

gunzip replied:
userName = fiscal Number

cloudify replied:
this is too confusing, `userName` is like my Twitter username, not a fiscal number

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+        type: "string"
+        description: "The user's fiscal number."
+        x-faker: "internet.userName"
+      emails: 
```
let's have only one email for now, I don't see any use core case that require / use multiple emails

gunzip replied:
multiple addres fields may be be useful when importing legacy data. moreover SCIM specs expect a multi-valued field.

cloudify replied:
`may be useful in the future` != keep it simple

gunzip replied:
i don't see "in the future" in my statement :)

cloudify replied:
I don't see any use case in the current use case list where we import multiple email addresses, nor how we could *use* multiple email addresses associated to a user, so I was assuming you had in mind a *future use case* - if you don't, then we're good 😄

gunzip replied:
Maybe. We probably have to think about an onboarding path for existing notifications services. 

Let's suppose [a service has already collected email from users](https://accessosicuro.rete.toscana.it/portal/accessError?targetSeviceUrl=https%3A%2F%2Fservizi.comune.campi-bisenzio.fi.it&errorCode=auth.access.error.message.noURLCertAuth&minRequiredAuthLevel=2&goToUrl=https%3A%2F%2Faccessosicuro.rete.toscana.it%3A443%2Fopensso%2FSSORedirect%2FmetaAlias%2Fidp%3FReqID%3D_b49def1e80d328f724bab2e525918c8d211f510b04%26spEntityId%3Dhttps%253A%252F%252Fservizi.comune.campi-bisenzio.fi.it%252F).

We can store these addresses as an email record associated with a service_id. When the service wants to send a message using the new API it will just provide the service_id (without the user email) and the message could be routed to this specific address (just brainstorming here).

cloudify commented on users.yaml:
```diff
@@ -0,0 +1,87 @@
+        description: "An email collection associated with a user. \n\nThere MUST be only one item marked primary."
+        items: 
+          $ref: "#/definitions/Email"
+      preferredLanguage: 
```
this should be an enum, with a list of supported locales, e.g:

* `it_IT`
* `it_CH`
* `en_GB`
* `en_US`
* `zh_CN`
...

also, may be worth having an array here, with multiple languages in order of preferences

gunzip replied:
an array here would break SCIM core schema compatibility. if we want to do that we have to create a new namespace for a custom multi-valued language field

cloudify replied:
let's not stick to SCIM then, it's making things more complicated, it's not solving any real problem we have

gunzip replied:
ok

cloudify commented:
@gunzip I agree with your comment about error codes, the only error code we need to document is 404 that is method-specific

cloudify commented:
Regarding the SCIM specs, I feel like we're trying to fit a very simple API into a complex spec, right now I honestly don't see the benefit of doing that, given that our API si very very simple and has its own requirements (i.e. having users identified by fiscalNumber instead of UUIDs).

Given the problems that fiscal numbers have (non uniqueness, and the fact that one's fiscal code may get blacklisted and/or modified) I see a future need of relying on a different user ID but, on the other end, I feel like overcomplicating the API.

@mccalv what's your opinion about identifying users by fiscal number?

gunzip commented:
About preserving the compatibility with SCIM specs I see some benefits:

1. There exist some good ready made implementations (free or commercial) that can be just plugged in from other Attribute Authorities and can be easily integrated with their already existing systems (API clients are ready made as well and the SCIM servers are interoperable): http://www.simplecloud.info/

2. It's widely used (Facebook, Oracle, CloudFoundry, Slack, Trello, GitHub, Microsoft) so developers can have some benefits by learning the API

3. It's compatible with SPID (= SAML) attributes: http://www.simplecloud.info/specs/draft-scim-saml2-binding-02.html

4. You can use it to store Oauth2 Client credentials and attributes (with the same API as the Users' ones)
  It's something we have to deal about and the risk is to reinvent the wheel:
  https://tools.ietf.org/html/draft-hunt-oauth-scim-client-reg-00

5. It has a nice publish / subscribe API (ietf draft currently): https://github.com/jirmauritz/scim-notification-hub to monitor attributes changes and sync the different systems.

cloudify commented:
I feel like all the benefits you're stating are either nice to haves or solutions to problems we may or we may not have in the future -> does for the principle of keeping it simple, I don't see how sticking to SCIM makes our life easier now

And by the way, in the future, if we see the actual need, we may always expose a SCIM compatible API wrapped around our simple user API.

So, I don't see the point of doing it now.

gunzip commented:
I somewhat disagree. We have some problems that we have to solve right now, like:

We are going to develop an open source profile authority that could be reused by other administrations to exposes users' custom attributes. It can be based on some documented open standard which has gained some traction or we can opt for our totally custom way. I'd rather go for the first route which will support interoperability with other systems.

Our API should be open to extensions from the beginning (and SCIM has a clear *predictable* path to extend resources schemas) that prevents a lot of breaking changes when you want to extend your domain models.

From the real beginning, we must have an API to deal with authorized Clients resources (organizations subscriptions, API keys, certs. etc.) If we limit the scope or our API to the bare minimum to handle users' preferences, we must design this new API from scratch. SCIM offers a way to seamlessy treat Client resources as Users resources.

cloudify commented:
> We are going to develop...

yes

> ...an open source...

yes

> ...profile authority... 

what's a profile authority? we're going to develop an API that can be used by citizens to provide well defined preferences related to digital contacts and language 

> ...that could be reused by other administrations... 

yes, hopefully

> ...to exposes users' custom attributes.

uhm no that's definitely not in the scope of this project, we're going to define a few attributes that we want to control - no custom attributes for now (we haven't defined any use case that require a custom attribute)

cloudify commented:
> Our API should be open to extensions from the beginning 

yes definitely

> (and SCIM has a clear predictable path to extend resources schemas) that prevents a lot of breaking changes when you want to extend your domain models.

yes indeed, the fact that SCIM *has* this capability, doesn't mean that SCIM is a right fit for our use cases, nor that a clear and predictable path to extend resources schemas [can't be implemented in other ways](https://zalando.github.io/restful-api-guidelines/compatibility/Compatibility.html)

The reasoning should be: we have this requirements, and this tools provides all of them and solves our problems with little or no side effect.

Instead here the reasoning is the opposite: we have this nice standard that everybody is using, let's change our use cases so that they fit nicely in the tool and also plan our strategy based on the feature that the tool provides 😸  

See the [Law of the instrument](https://en.wikipedia.org/wiki/Law_of_the_instrument)

cloudify commented:
> From the real beginning, we must have an API to deal with authorized Clients resources (organizations subscriptions, API keys, certs. etc.) If we limit the scope or our API to the bare minimum to handle users' preferences, we must design this new API from scratch. SCIM offers a way to seamlessy treat Client resources as Users resources.

I don't understand the topic here, the scope of this API is to be the simplest, yet useful API for handling user preferences, that's what we want to provide to the early adopters of this API - why we need to deal with organizations subscriptions, API keys, certs ?!?

gunzip commented:
All considerations about diverging from SCIM specs have been accepted and are now reflected in the YAML files.

As a compromise we can treat user's preferences as a user *subresource* so we leave the `/User/*` endpoints "free" for future extensions.

I still favor the multi-value email address field which IMHO is more future proof (it's easier to refactor from an object array to a single- field than vice versa) and could be useful to import legacy data as stated above.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+        '200':
+          description: Message found.
+          schema:
+            $ref: '#/definitions/AcceptedMessage'
```
submitted message -> `SubmittedMessage`

or

accepted message -> `AcceptedMessage`

We should use the same terminology everywhere (in the model names, in the descriptions, in the state enum etc...)

gunzip replied:
i propose to use SubmittedMessage for input model and AcceptedMessage for output model (with status and id)

cloudify replied:
ah, I see where the confusion comes from, if you name it `SubmittedMessage`, for me that's the state of the model after the fact that is was submitted (past tense), instead what you mean is when the object is about to be submitted (before the event) - then I guess you should find a different naming, something like `NewMessage`

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+        in: path
+        type: string
+        required: true
+        description: The fiscal number of the user.
```
The fiscal number of the user, all upper case.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+        pattern: >-
+          [A-Z]{6}[0-9LMNPQRSTUV]{2}[ABCDEHLMPRST][0-9LMNPQRSTUV]{2}[A-Z][0-9LMNPQRSTUV]{3}[A-Z]
+        maxLength: 16
+        minLength: 1
```
why minLength = 1? fiscal code [is always 16 characters](https://it.wikipedia.org/wiki/Codice_fiscale#Generazione_del_codice_fiscale) for a person

gunzip replied:
right

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+      operationId: getMessagesByUser
+      summary: Get messages by user
+      description: |-
+        Returns the submitted messages the have the user as destination.
```
the have -> that has

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+      summary: Get messages by user
+      description: |-
+        Returns the submitted messages the have the user as destination.
+        Messages will be returned in inverse delivery order (from last to first)
```
inverse submission order

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+              - $ref: '#/definitions/PaginationResponse'
+          examples: {}
+        '404':
+          description: No user found.
```
No messages have been submitted for the user identified by the provided fiscal code.

gunzip replied:
this is really 'no user found' (with the provided fiscal code).

when no message are submitted there's a 200OK with empty  items: [] array

cloudify replied:
how would you tell the difference?

cloudify replied:
I thin the correct answer for the absence of messages of a user is a 404, since the request if for the `Message` resource, not the `User` resource.

Suppose the user X have no associated messages, would you answer to the question "Please give me all messages sent to user X", with:

* OK (200), I've found messages, here's the list of messages: `[ empty list ]`

or

* NOT FOUND (404), Sorry I haven't found any messages for that user

?

gunzip replied:
There are no standard and/or common view about it:
https://www.google.it/search?q=http+status+code+for+empty+result

I prefer the 200OK semantic for empty collections as stating that a collection is empty IMHO differs from stating that the collection does not exists at all (because there's no user associated).

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+        Messages will be returned in inverse delivery order (from last to first)
+      responses:
+        '200':
+          description: Found.
```
Returns the messages for the user identified by the provided fiscal code.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+        '404':
+          description: No user found.
+      parameters: []
+  /messages:
```
I wonder if we should, for consistency with the getMessage method, have the `messages` resource nested under `user`, i.e. `POST /users/{fiscal_number}/messages`

this could also have an additional benefit when distributing the load to multiple shards

gunzip replied:
probably we should.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+            Location:
+              type: string
+              description: >-
+                Location (URL) of accepted message resource.
```
accepted -> submitted ?

gunzip replied:
I think "accepted" is more correct here as submitted messages can be rejected. Moreover, if we call the input model 'SubmittedMessage' and the output model 'AcceptedMessage' this rule is enforced.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+    properties:
+      body_short:
+        type: string
+        description: Message content (short version).
```
An abstract of the message, in plain text. The content of this field will be delivered to channels that support a limited amount of characters (e.g. SMS, mobile notifications, etc...).

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+        maxLength: 100
+      body_long:
+        type: string
+        description: Message content (long version).
```
A long version of the message, in plain text. The content of this field will be delivered to channels that don't have any limit in terms of content size (e.g. email, etc...). The long version of the message is optional and should be provided only in cases when a longer message (compared to the short version provided in `body_short`) gives more information to the user. This field should not be a duplicate of `body_short`.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+      body_long:
+        type: string
+        description: Message content (long version).
+        minLength: 3
```
given that this should not be a duplicate of `body_short` but should be longer, I would say it should be > 100

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+    type: object
+    properties:
+      to:
+        description: The fiscal code of the destination user.
```
upper case

can we have a pattern here too?

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+      dry_run:
+        type: boolean
+        description: >-
+          This parameter, when set to true, allows developers to test a request
```
more specific here about the semantic, will the caller get back a UUID of the message? what's the status of the message going to be?

gunzip replied:
actually I have no idea. I wonder the same about firebase cloud messaging API (what's a "dry run" in their scenario ?). what's a dry run useful for ? to validate the passed input message ? in this case we can return 

- 200 OK with empty payload for valid messages (without storing anything)
- 500 + problem+json in the payload with a validation error in the other case

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+      time_to_live:
+        type: integer
+        description: >-
+          This parameter specifies how long (in seconds) the message should be
```
This parameter specifies for how long (in seconds) the system will try to deliver the message to the channels configured by the user.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+      status:
+        type: string
+        description: |
+          ACCEPTED: the message has been accepted for delivery
```
then we should always use *accepted* instead of *submitted*

although, I fear that *accepted* may be misleading, as one may think it has been *accepted* by the user (?)

gunzip replied:
i think that accepted (by the system) != submitted (by the sender, can be rejected)

cloudify replied:
what exactly you mean by *rejected*?

gunzip replied:
ie.  the provided payload (POST body) failed validation for some fields or the system failed to queued the message.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+  MessageStatus:
+    title: MessageStatus
+    type: object
+    properties:
```
no `accepted_at` ?

gunzip replied:
it's in the AcceptedMessage payload now. it depends on the samantic: we can have a global 'accepted_at' which means 'accepted by the whole system' or 'accepted_at' here, in message status, which depends on the specific channel (so multiple 'accepted_at' one for each channel).

cloudify replied:
I see, the confusion for me is that the `ACCEPTED` state is provided for each channel, but the `accepted_at` is for the whole message - something is semantically not right

also another issue I'm thinking of is: why the sender, that cannot decide what channels the message should be sent to, may want to know the status of delivery for each channel? and actually has to check the status for each channel? at the end what he wants to know is whether the message has been delivered to the user through at least a channel (the fact that may be more than one channel is actually purposely hidden from him)

so I think that we should not expose the state of each channel, but just tell the sender whether the message as been delivered to at least one channel

also, how is the information that some channel succeeded but some other channel failed useful to the sender? if he attempts to retry the delivery by sending a new message, given that he cannot limit the channels to the failed ones, he would send duplicate messages!!!

so I don't see how is actually useful to provide the status information for each channel

gunzip replied:
probably you're right and we can collapse status. however this exposes new issues we never considered before ie:

> why the sender, that cannot decide what channels the message should be sent to, may want to know the status of delivery for each channel ?

because when she sends `body_short` and `body_long` (et al.) wants to know which part of the payload has been delivered (ie. if you just return the ) and not all channels support all payload types. (in this case probably it's better to have a status for each payload type and abstract the channel away :)

> but just tell the sender whether the message as been delivered to at least one channel

> how is the information that some channel succeeded but some other channel failed useful to the sender?

same as above. you lose the info on which part of the payload gets delivered.

cloudify replied:
What can the client do with that information? `body_short` should contain the essence of the message, it doesn't make sense to have a short version if it's missing any important piece of information - the message should be able to exist solely in the short version - the longer version is there just as an "improved representation" of the message. The caller should assume that only the `body_short` version may be delivered and that version should contain all the required information. We should make this very clear and explicit in the docs (and check wether the early adopters understand this requirement from the documentation).

In fact `body_short` is required and `body_long` is optional.

gunzip replied:
ok

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+          id:
+            type: string
+            format: uuid
+          accepted_at:
```
I would say for consistency this should go along with `failed_at` and `sent_at`

gunzip replied:
see comment above. these are per channel values.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+                $ref: '#/definitions/MessageStatus'
+              browser:
+                $ref: '#/definitions/MessageStatus'
+              push:
```
I would call this `app` as these are all kind of push :)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+            default: {}
+            properties:
+              mail:
+                $ref: '#/definitions/MessageStatus'
```
what status will be the message when the channel is not enabled?

gunzip replied:
rejected ?

cloudify replied:
you mean `FAILED`? I don't see a `rejected` status

gunzip replied:
yes FAILED. according to what we have discussed above, now we have a global FAILED status jsut in case *every* channel fails (is it right ?)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+  PaginationResponse:
+    type: object
+    title: pagination
+    description: Pagination response parameters (HATEOAS version).
```
I'd leave `(HATEOAS version)` out of the description, doesn't provide any useful information to the user of the API.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+    title: pagination
+    description: Pagination response parameters (HATEOAS version).
+    properties:
+      index:
```
what's the use of `index` ?

gunzip replied:
it's an offset (taken straight from http://zalando.github.io/restful-api-guidelines/pagination/Pagination.html)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+        type: integer
+        minimum: 1
+        description: Number of items returned for each page.
+      first:
```
I would leave `first`, `prev`, `last` and `self` out for now

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,268 @@
+    in: query
+    type: string
+    minimum: 1
+    description: An opaque identifier that points to the next item in the  collection.
```
there's an extra space at `the  collection`

cloudify commented on preferences.yaml:
```diff
@@ -0,0 +1,76 @@
+schemes:
+  - https
+paths:
+  '/users/{fiscal_number}/preferences':
```
why not just `/users/{fiscal_number}` ?

`preferences` is not an actual resource since you cannot access each `preference` independently, if you had say `GET /users/{fiscal_number}/preferences/{preference_id}` then it would make sense, in this case the preferences are just attribute of the `User` resource

cloudify replied:
or perhaps if you want to talk about Preferences (as a collection of attributes) instead of Users you can have `/preferences/{fiscal_number}`

gunzip replied:
GET /users/{fiscal_number}/preferences/{preference_id} would be useful in the case of resource versioning. We can now assume that "/users/{fiscal_number}/preferences" always points to the latest revision. Moreover a user it's not his preferences so the '/User' endpoint should be free IMHO. Said that, it's ok for me to use "/preferences/{fiscal_number}". Should we settle on it?

gunzip replied:
just want to point out that for messages we use '/user/{}/messages' anyway, so maybe '/user/{}/preferences' would be consistent.

cloudify commented on preferences.yaml:
```diff
@@ -0,0 +1,76 @@
+        '200':
+          description: Found.
+          schema:
+            $ref: '#/definitions/User'
```
so, if you're returning a `User` then it should be `/users/{fiscal_number}`

just make it consistent :)

gunzip replied:
right, I'll change it to Preferences

cloudify commented on preferences.yaml:
```diff
@@ -0,0 +1,76 @@
+    get:
+      operationId: getUserPreferences
+      summary: Get User's Preferences
+      description: Retrieve one user's preferences (ie. email and/or preferred language).
```
Returns the preferences for the user identified by the provided fiscal number.

cloudify commented on preferences.yaml:
```diff
@@ -0,0 +1,76 @@
+            $ref: '#/definitions/User'
+          examples: {}
+          headers:
+            ETag:
```
this (in tandem with `If-Match`) is useful only to avoid update conflicts - since this spec doesn't cover updates, what's the use of ETags here?

cloudify commented on preferences.yaml:
```diff
@@ -0,0 +1,76 @@
+      preferred_languages:
+        type: array
+        description: >-
+          Indicates the User's preferred written or spoken language. Generally
```
this is too much, not all combinations are valid / make sense, to make the life of the client simpler I would use an enum here, so they know what languages they should try to cover. I would start with:

* `it_IT`
* `en_GB`
* `es_ES`

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,275 @@
+          (compared to the short version provided in body_short) gives more
+          information to the user. This field should not be a duplicate of
+          body_short.
+        minLength: 100
```
it's sane to still have a reasonable maxLength

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,275 @@
+      dry_run:
+        type: boolean
+        description: >-
+          This parameter, when set to true, allows developers to test a request
```
again, needs better description about what to expect as response (do I still get a 201 with the UUID of the message created? will I be able to issue a get ?)

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,275 @@
+    title: pagination
+    description: Pagination response parameters.
+    properties:
+      index:
```
how would you use this `index`? I don't understand

cloudify replied:
Also there should be an explanation of how pagination works on the description of the method that returns paginated results

gunzip commented:
I'm closing this PR since GitHub has some hiccups above > 200 comments and the page take > 10sec to open :) Moreover a lot has changed from the beginning, so we'd better start with a new clean state.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,242 @@
+        type: string
+        format: date-time
+        description: 'The date-time the message has been created (delivered to the user''s inbox). The format is defined by RFC 3339, for example 2015-05-28T14:07:17Z.'
+      read_at:
```
This should be optional (can be null in case the user has not yet read the message), not sure how you would define that in the specs, at least we should explain it in the description.

Also, we should explain that this has set-once behavior.

cloudify commented on notifications.yaml:
```diff
@@ -0,0 +1,242 @@
+    properties:
+      id:
+        type: string
+        format: uuid
```
this has to be a string, we're going to use [ULIDs](https://github.com/alizain/ulid) that are lexicographically sortable and makes it easier to naturally sort the messages by creation time.

gunzip replied:
ok

cloudify commented on preferences.yaml:
```diff
@@ -0,0 +1,67 @@
+schemes:
+  - https
+paths:
+  '/users/{fiscal_code}/preferences':
```
`preferences` is not a nested resource, compare this to `/users/{fiscal_code}/messages`

gunzip replied:
ok

gunzip commented on notifications.yaml:
```diff
@@ -0,0 +1,242 @@
+          examples: {}
+        '404':
+          description: No message found for the provided ID.
+  '/users/{fiscal_code}/messages':
```
sorry, can you pls elaborate more ? (what's for internal use exactly?)

cloudify replied:
Let's merge this PR now and extract this part in another PR

gunzip commented on notifications.yaml:
```diff
@@ -0,0 +1,242 @@
+                  items:
+                    type: array
+                    items:
+                      $ref: '#/definitions/CreatedMessage'
```
> For this reason

you mean just to have a shorter payload in the response ? 
maybe it's worth to have at least body_short to avoid a N+1 select problem.

cloudify replied:
ok let's leave this discussion for a separate PR, this one is becoming too big!
